File: grind75/week2/014-Binary-Tree-Right-Side-View-Medium/014-Binary-Tree-Right-Side-View-Medium.py
# problem link: https://leetcode.com/problems/binary-tree-right-side-view/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:

        # base case
        if not root:
            return []

        queue = collections.deque([root])
        res = []
        while queue:
            level_len = len(queue)

            for i in range(level_len):
                node = queue.popleft()
                
                if i == level_len - 1: #right mode node
                    res.append(node.val)
                
                if node.left:
                    queue.append(node.left)
                
                if node.right:
                    queue.append(node.right)
                
        return res
        
        

File: grind75/week2/014-Binary-Tree-Right-Side-View-Medium/014-Binary-Tree-Right-Side-View-Medium.cpp
// link: https://leetcode.com/problems/binary-tree-right-side-view/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> ans;

        // edge case
        if (!root)
            return ans;

        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty()) {
            int level_len = q.size();

            for (int i = 0; i < level_len; i++) 
            {
                TreeNode* node = q.front();
                q.pop();

                if (node->left)
                    q.push(node->left);
                
                if (node->right)
                    q.push(node->right);

                if (i == level_len - 1) // rightmost node at this level
                    ans.push_back(node->val);
            }
        }

        return ans;
    }
};
